src/main/java/baritone/command/defaults/CraftCommand.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * This file is part of Baritone.
+ *
+ * Baritone is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Baritone is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Baritone.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package baritone.command.defaults;
+
+import baritone.api.IBaritone;
+import baritone.api.command.Command;
+import baritone.api.command.argument.IArgConsumer;
+import baritone.api.command.datatypes.ItemById;
+import baritone.api.command.exception.CommandException;
+import net.minecraft.item.Item;
+import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.item.crafting.IRecipe;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+
+public class CraftCommand extends Command {
+
+    protected CraftCommand(IBaritone baritone) {
+        super(baritone, "craft");
+    }
+
+    @Override
+    public void execute(String label, IArgConsumer args) throws CommandException {
+        String itemName = args.getString();
+        Item item = Item.getByNameOrId(itemName);
+
+        int amount = args.hasAny() ? args.getAs(Integer.class) : 1;
+
+        if (item == null) {//this is hacky, but it gets the job done for now. also we arnt interested in non craft-able items anyway
+            itemName = itemName.replace("_", " ");
```

Comment:
i understand the logic behind the first comment but not the second.

Reply:
Basically I mean you should first get the item by either method (id or name) and then continue along the same code path for both cases.
Currently this is
```
get item by id
if that failed
      get by name
     handle all the resource counting
else
     handle all the resource counting slightly differently
```
and I think it should be
```
get item by id
if that failed
    get item by name
handle all the resource counting (the same for both)
```
Why should there be a difference between `#craft wooden_sword` and `#craft wooden sword`?
Not sure how easy that is though, due to the way you use items in the id case and recipes in the name case.

Reply:
it maybe doesnt make that much sense with a wooden sword as there only exist 1 wooden sword.
take the example of a red bed

if you type: `#craft bed` it will craft any bed this can be a white bed, a blue bed or a red bed.
if you type `#craft red bed` it wont find a item that is named like that so it searches a recipe that produces a item stack with that display name. note here that there are 2 recipes for red beds, one with red wool and one dying a white bed.

i use crafItem() if i want any instance of that item crafted and i use craftRecipe() if i want baritone to use that specific recipe to craft

Reply:
Can't you get all recipes that produce the desired item (be it by name or id) and then use that collection of recipes from that point on? That way there is only one case (n recipes) instead of two. In general even if the user specified one exact item there can still be multiple recipes for it and the current recipe switching mechanism is sketchy (doesn't it even bypass what you said about `#craft red bed` because `redBedRecipe.getRecipeOutput().getItem() == Items.BED` and getting a recipe for that doesn't need to return `redBedRecipe` again?).